src/ucai/core/utils/callable_utils.py

```diff
@@ -0,0 +1,142 @@
+import ast
+import inspect
+from textwrap import dedent
+from typing import Any, Callable, get_type_hints
+
+from ucai.core.utils.type_utils import python_type_to_sql_type, validate_container_type
+
+FORBIDDEN_PARAMS = ['self', 'cls']
+
+def extract_function_body(func: Callable[..., Any]) -> str:
+    """Extracts the body of a function as a string without the signature or docstring, preserving indentation."""
+    parsed_source, source_lines = parse_dedented_source(func)
+    
+    function_def = parsed_source.body[0]
+
+    body_start_line = function_def.body[0].lineno - 1
+    
+    if isinstance(function_def.body[0], ast.Expr) and isinstance(function_def.body[0].value, ast.Constant):
+        body_start_line += len(function_def.body[0].value.s.splitlines())
+
+    function_body_lines = source_lines[body_start_line:]
+
+    min_indent = min((len(line) - len(line.lstrip()) for line in function_body_lines if line.strip()), default=0)
+    
+    function_body = "\n".join([line[min_indent:] if len(line) >= min_indent else line for line in function_body_lines])
+
+    return function_body
+
+def parse_docstring(docstring: str) -> dict:
```

Comment:
We can use a state machine like approach to improve the code.

```python
from dataclasses import dataclass
from typing import Dict, List, Optional

@dataclass
class ParsedDocstring:
    args: Dict[str, str]
    returns: Optional[str] = None
    raises: Optional[str] = None

def parse_docstring(docstring: str) -> ParsedDocstring:
    """Parses the docstring to extract comments for parameters, return value, and exceptions raised."""
    parsed_docstring = ParsedDocstring(args={})
    state: Optional[str] = None  # Possible states: 'args', 'returns', 'raises'
    current_param: Optional[str] = None
    description_lines: List[str] = []

    for line in docstring.splitlines():
        line = line.strip()
        
        if line.startswith("Args:"):
            state = 'args'
            continue
        elif line.startswith("Returns:"):
            state = 'returns'
            continue
        elif line.startswith("Raises:"):
            state = 'raises'
            continue

        if state == 'args':
            if line and ":" in line:
                if current_param and description_lines:
                    parsed_docstring.args[current_param] = " ".join(description_lines).strip()
                
                param_part, comment_part = line.split(":", 1)
                current_param = param_part.split("(")[0].strip()
                description_lines = [comment_part.strip()]
            elif current_param:
                description_lines.append(line)

        elif state == 'returns':
            parsed_docstring.returns = line.strip()

        elif state == 'raises':
            if line:
                parsed_docstring.raises = line.strip()

    # Finalize last parameter if in 'args' state
    if state == 'args' and current_param and description_lines:
        parsed_docstring.args[current_param] = " ".join(description_lines).strip()

    return parsed_docstring

```

Reply:
Interesting idea! Adjusted the suggestion to fit within the design :)